db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ValueNode.java

```diff
@@ -1036,10 +1036,24 @@ public double joinSelectivity(Optimizable optTable,
                 selectivity = 1.0d/innerRowCount;
                 break;
             case INNER:
-                selectivity = (1.0d-.1d)*(1.0d-.1d)/Math.min(innerRowCount,outerRowCount);
-                break;
-            case ANTIJOIN:
-                selectivity = 1.0d-((1.0d-.1d)*(1.0d-.1d)/Math.min(innerRowCount,outerRowCount));
+			case ANTIJOIN:
+                selectivity = (1.0d-.1d)*(1.0d-.1d)/Math.max(innerRowCount,outerRowCount);
+                if ((optTable instanceof FromTable && ((FromTable) optTable).getExistsTable())) {
+					if (outerRowCount >= innerRowCount) {
+						selectivity = (1 - .1d) / outerRowCount;
+						if (selectivityJoinType == SelectivityUtil.SelectivityJoinType.ANTIJOIN) {
+							// output row = left not null row - output row estimated above for inclusion join
+							selectivity = (1 - .1d) / outerRowCount * (outerRowCount - innerRowCount) / innerRowCount;
+						}
+					} else {
+						selectivity = (1 - .1d)/innerRowCount;
+						if (selectivityJoinType == SelectivityUtil.SelectivityJoinType.ANTIJOIN) {
+							// following the above assumption, no row will be returned, but to be conservative, we will
+							// assume 10% selectivity from left rows
+							selectivity = 0.1 * selectivity;
+						}
+					}
+				}
```

Comment:
There are 3 scenarios for INNER JOIN: regular inner join, inclusion join and exclusion join. For legacy reason, the exclusion join is given a separate enum value ANTIJOIN, but not the inclusion join. The original code uses the same costing logic for regular INNER JOIN and inclusion join, this is wrong. The fix takes into consideration all 3 scenarios but since the inclusion and exclusion logic has some overlapping, I put all of them in one block. 
This is also consistent with the logic in BinaryRelationalOperatorNode.joinSelectivity().

Reply:
That makes sense, but I find the code hard to read because this double case clearly isn't a double case anymore as there is a lot of dead stores depending on the branching. Therefore the suggestion I made.

Reply:
Thanks, I've restructured the code.